src/main/scala/firrtl/passes/UpdateDuplicateMemMacros.scala

```diff
@@ -0,0 +1,95 @@
+package firrtl.passes
+
+import scala.collection.mutable.{HashMap,ArrayBuffer}
+import AnalysisUtils._
+import MemTransformUtils._
+import firrtl.ir._
+import firrtl._
+import firrtl.Mappers._
+import firrtl.Utils._
+
+object MemTransformUtils {
+
+  def createRef(n: String) = WRef(n,UnknownType,ExpKind(),UNKNOWNGENDER)
+  def createSubField(exp: Expression, n: String) = WSubField(exp,n,UnknownType,UNKNOWNGENDER)
+  def connectFields(lref: Expression, lname: String, rref: Expression, rname: String) = 
+    Connect(NoInfo,createSubField(lref,lname),createSubField(rref,rname))
+
+  def getMemPortMap(m: DefMemory) = {
+    val memPortMap = HashMap[String,Expression]()
+    val defaultFields = Seq("addr","en","clk")
+    val rFields = defaultFields :+ "data"
+    val wFields = rFields :+ "mask"
+    val rwFields = defaultFields ++ Seq("wmode","wdata","rdata","wmask")
+
+    def updateMemPortMap(ports: Seq[String], fields: Seq[String], portType: String) = 
+      for (p <- ports.zipWithIndex; f <- fields) {
+        val newPort = createSubField(createRef(m.name),portType+p._2)        
+        val field = createSubField(newPort,f)
+        memPortMap(s"${m.name}.${p._1}.${f}") = field
+      }
```

Comment:
I generally prefer unapplying tuples to the the whole ._1 and ._2. ie.
```scala
for ((p, i) <- ports.zipWithIndex; f <- fields) {
```
then you can use i for the index and p for the port!